### PR TITLE
feat(Providers): Add more options for injecting provider values

### LINKED ISSUE #23: Request: additional options for injecting provider values
Angular DI offers several options for providing dependencies to the container:

```ts
[
  { provide: ..., useClass: FakeLogger },
  { provide: ..., useExisting: Logger},
  { provide: ..., useValue: { log: () => {} } },
  { provide: ..., useFactory: createLogger }
]
```

In particular, it would be great if we could provide a `factory` function. I think this would be much better than the existing solution that requires you use `@Service({ factory: ... })`. With the factory defined in the service decorator it's much more difficult to mock out in tests.

```ts
Container.provide([
  { name: 'foo', value: new Foo() },
  { name: 'foo.class', class: Foo },
  { name: 'foo.factory', factory: () => new Foo() },
])
```

### PR DESCRIPTION
**Pull request by xxxtonixxx**
Closes #23

### CODE REVIEW

**Comment by xxxtonixxx** [comment]
friendly ping @pleerock. Could you review it?, I would like to see this in typedi 👍

**Comment by pleerock** [comment]
@xxxtonixxx sorry for late response

**Comment by pleerock** [comment]
did you see `registerService` function and particular its ability to accept a factory function? I don't understand why your PR is so huge, there shouldnt be so much changes to implement this requested feature since all implementation is there

**Comment by xxxtonixxx** [comment]
Yes, I saw it and [it's used](https://github.com/pleerock/typedi/pull/35/files#diff-448ab686459f478b589cd097da39a6e6R215). However, it improves the API and

>  offers several options for providing dependencies

I have added some user-friendly features, like:
-  `Container.provide([ClassOne, ClassTwo])`
- `Container.provide([{ id: "factory", factory: (injectedString, injectedNumber, classOne) => {...}, deps: ["tokenOne", tokenTwo, ClassOne] }])`
  - Before, you could not inject strings, numbers or whatever you wanted
- Container.provide([{ id: "myAlias", aliasOf: "tokenOne" }])
- And `Container.provide` receives an array.

It could be implemented over `registerService`, however IMO `provide` method is cleaner than `registerService`, although, `provide` method uses `registerService` under the hat. And it doesn't break compatibility.

So, almost 50% of code of PR are tests and typings of `providers`.

**Comment by pleerock** [comment]
My point was to take a look on `registerService` and see that it supports a factory method already.

> So, almost 50% of code of PR are tests and typings of providers.

yeah I was talking excluding tests in mind

File: src/Container.ts [hunk 1]
```diff
@@ -69,6 +72,10 @@ export class Container {
         // find if service already was registered
         let service = this.findService(identifier);
 
+        if (service && service.aliasOf) {
```
Comment (pleerock) [inline src/Container.ts]: add a comment what are u doing here before this line

File: src/Container.ts [hunk 2]
```diff
@@ -89,12 +96,25 @@ export class Container {
...
+            // setup constructor parameters for a newly initialized service
+            const paramTypes = Reflect && (Reflect as any).getMetadata ? (Reflect as any).getMetadata("design:paramtypes", type) : undefined;
+            params = paramTypes ? this.initializeParams(type, paramTypes) : [];
+        }
 
         // if factory is set then use it to create service instance
         if (service.factory) {
+            
+            // fill the params array with the deps defined
+            if (service.deps) {
```
Comment (pleerock) [inline src/Container.ts]: "deps" maybe better call it "dependencies"?

File: src/Container.ts [hunk 3]
```diff
@@ -162,8 +182,45 @@ export class Container {
...
+            // if it is a function/class, it can be registered
+            if (this.isFunction(provider)) {
+                this.registerService({
+                    type: provider,
+                });
+
+                return;
+            }
+
+            if (provider.id == null) {
```
Comment (pleerock) [inline src/Container.ts]: use strict checking, e.g. ===

File: src/Container.ts [hunk 4]
```diff
@@ -162,8 +182,45 @@ export class Container {
...
      * Provides a set of values to be saved in the container.
      */
-    static provide(values: { id: ServiceIdentifier, value: any }[]) {
-        values.forEach((v: any) => this.set(v.id, v.value));
+    static provide(providers: (Provider | ObjectType<any>)[]) {
+        providers.forEach(provider => {
+            // if it is a function/class, it can be registered
+            if (this.isFunction(provider)) {
+                this.registerService({
+                    type: provider,
```
Comment (pleerock) [inline src/Container.ts]: what is the point to register service without a value?

File: src/types/Providers.ts [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+export interface ProviderByID {
+    id: ServiceIdentifier;
+}
+
+/**
+ * Use a class as service.
+ * It's similiar to add directly the class.
+ * Container.provide([TestService]) === Container.Provider([{id: TestService, class: TestService}])
+ */
+export interface ClassProvider extends ProviderByID {
```
Comment (pleerock) [inline src/types/Providers.ts]: same question here - why do we need to provide a class without a value?

File: src/types/Providers.ts [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+export interface FactoryProvider extends ProviderByID {
+    factory: [ObjectType<any>, any]|((...params: any[]) => any);
+    deps?: ServiceIdentifier[];
+}
+
+/**
+ * Use an already defined service
+ */
+export interface ExistingProvider extends ProviderByID {
+    existing: ServiceIdentifier;
```
Comment (pleerock) [inline src/types/Providers.ts]: is it same as "aliasOf" ? If yes, then maybe call it same - "aliasOf"?

File: src/types/ServiceMetadata.ts [hunk 7]
```diff
@@ -30,4 +31,13 @@ export interface ServiceMetadata<T, K extends keyof T> {
      */
     instance?: Object;
 
+    /**
+     *
```
Comment (pleerock) [inline src/types/ServiceMetadata.ts]: place good comments

File: src/types/ServiceMetadata.ts [hunk 8]
```diff
@@ -30,4 +31,13 @@ export interface ServiceMetadata<T, K extends keyof T> {
      */
     instance?: Object;
 
+    /**
+     * 
+     */
+    aliasOf?: ServiceIdentifier;
+
+    /**
+     *
```
Comment (pleerock) [inline src/types/ServiceMetadata.ts]: place good comments

File: src/types/ServiceMetadata.ts [hunk 9]
```diff
@@ -30,4 +31,13 @@ export interface ServiceMetadata<T, K extends keyof T> {
...
 
+    /**
+     * 
+     */
+    aliasOf?: ServiceIdentifier;
+
+    /**
+     * 
+     */
+    deps?: ServiceIdentifier[];
```
Comment (pleerock) [inline src/types/ServiceMetadata.ts]: probably better to call it "dependencies"

File: src/Container.ts [hunk 10]
```diff
@@ -162,8 +182,45 @@ export class Container {
     /**
      * Provides a set of values to be saved in the container.
      */
-    static provide(values: { id: ServiceIdentifier, value: any }[]) {
-        values.forEach((v: any) => this.set(v.id, v.value));
+    static provide(providers: (Provider | ObjectType<any>)[]) {
```
Comment (pleerock) [inline src/Container.ts]: idea of "provide" was to make it work same as "set", but for multiple items. But right now "set" does not support everything provide supports. Maybe move some peaces of code into "set" method and here simply call it?

Comment (pleerock) [reply src/Container.ts]: also we can remove `registerHandler` function and merge its functionality into `set` method:

```ts

    /**
     * Sets a value for the given type or service name in the container.
     */
    static set<T, K extends keyof T>(service: ServiceMetadata<T, K>): Container;

    /**
     * Sets a value for the given type or service name in the container.
     */
    static set(type: Function, value: any): Container;

    /**
     * Sets a value for the given type or service name in the container.
     */
    static set(name: string, value: any): Container;

    /**
     * Sets a value for the given type or service name in the container.
     */
    static set(token: Token<any>, value: any): Container;

    /**
     * Sets a value for the given type or service name in the container.
     */
    static set(identifierOrServiceMetadata: ServiceIdentifier|ServiceMetadata<any, any>, value?: any): Container {

        const newService: ServiceMetadata<any, any> = arguments.length === 1 && identifierOrServiceMetadata instanceof Object ? identifierOrServiceMetadata : undefined;
        if (newService) {
            const service = this.findService(newService.id);
            if (service) {
                Object.assign(service, newService);
            } else {
                this.services.push(newService);
            }

        } else {
            const identifier = identifierOrServiceMetadata as ServiceIdentifier;
            const service = this.findService(identifier);
            if (service) {
                service.instance = value;

            } else {
                const service: ServiceMetadata<any, any> = {
                    instance: value
                };
                if (identifier instanceof Token || typeof identifier === "string") {
                    service.id = identifier;

                } else {
                    service.type = identifier;
                }

                this.services.push(service);
            }
        }

        return this;
    }
```

File: test/Container.spec.ts [hunk 11]
```diff
@@ -101,6 +104,189 @@ describe("Container", function() {
 
         });
 
+        describe("should be able to set a", () => {
```
Comment (pleerock) [inline test/Container.spec.ts]: set what?

**Comment by pleerock** [comment]
any updates on this PR @xxxtonixxx ?

**Comment by pleerock** [comment]
Ive manually applied some your changes and refactored some of the parts I didn't liked. Thank you very much for your efforts and contribution!

**Pull request closed without merging**